aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -916,6 +917,142 @@ Tensor norm(const Tensor& self, optional<Scalar> p, DimnameList dim, bool keepdi
+    "Expected dims to be different, got (", dim[0], ", ", dim[1], ") instead");
+
+  Tensor result_;
+  Tensor self_;
+
+  if (opt_dtype.has_value()) {
+    self_ = self.to(scalarType);
+  } else {
+    self_ = self;
+  }
+  TORCH_CHECK(result.dtype() == self_.dtype(),
+    "_norm_matrix: provided dtype must match dtype of result. Got ",
+    result.dtype(), " and ", self_.dtype());
+
+  if (std::abs(p) == 2) {
+    // Need to shift the reduction dims to the back, because at::svd will only operate on
+    // the last 2 dimensions
+    auto permutation = create_dim_backshift_permutation(dim_[0], dim_[1], self.dim());
+    auto permutation_reverse = create_reverse_permutation(permutation);
+
+    result_ = std::get<1>(self_.permute(permutation).svd()).abs();
+
+    if (p > 0) {
+      result_ = std::get<0>(result_.max(-1, keepdim));
+    } else {
+      result_ = std::get<0>(result_.min(-1, keepdim));
+    }
+
+    if (keepdim) {
+      result_.unsqueeze_(-1);
+      result_ = result_.permute(permutation_reverse);
+    }
+  } else {
+    // abs(p) == infinity and abs(p) == 1 will perform identical reductions, except
+    // that the order of the two dims is swapped. So we can swap the dims if
+    // abs(p) == infinity to simplify the rest of the operation's logic.
+    if (std::abs(p) == INFINITY) {
+      std::swap(dim_[0], dim_[1]);
+    }
+    // If the dim of the second reduction is greater than that of the first reduction
+    // and we are not keeping the dims, then the fact that the output of the first
+    // reduction will have one fewer dimension means that the second reduction dim
+    // will be off by one, so we need to correct that.
+    if ((dim_[1] > dim_[0]) && !keepdim) {
+      dim_[1]--;
+    }
+    if (p == 1 || p == INFINITY) {
+      result_ = std::get<0>(self_.abs().sum(dim_[0], keepdim).max(dim_[1], keepdim));
+    } else if (p == -1 || p == -INFINITY) {
+      result_ = std::get<0>(self_.abs().sum(dim_[0], keepdim).min(dim_[1], keepdim));
+    } else {
+      TORCH_CHECK(false, "Order ", p, " not supported for matrix norm");
+    }
+  }
+  result.resize_as_(result_);
+  result.copy_(result_);
+  return result;
+}
+
+Tensor _norm_matrix(const Tensor &self, Scalar p) {
+  Tensor result = at::empty_like(self);
+  return at::native::_norm_matrix_out(result, self, p, IntArrayRef{}, false, c10::nullopt);
+}
+
+Tensor &_norm_matrix_out(Tensor& result, const Tensor& self, optional<Scalar> p, IntArrayRef dim, bool keepdim, ScalarType dtype) {
+  return at::native::_norm_matrix_out(result, self, p, dim, keepdim, optional<ScalarType>(dtype));
+}
+
+Tensor &_norm_matrix_out(Tensor& result, const Tensor& self, optional<Scalar> p, IntArrayRef dim, bool keepdim) {
+  return at::native::_norm_matrix_out(result, self, p, dim, keepdim, c10::nullopt);
+}
+
+static Tensor _norm_matrix(const Tensor& self, optional<Scalar> p, IntArrayRef dim, bool keepdim,
+            optional<ScalarType> opt_dtype) {
+  Tensor result = at::empty_like(self);
```

Comment:
Do you to size or specify the type of the result here? (vs. just passing in `Tensor result`)?

Reply:
I don't necessarily have to specify the size, but I do have to specify the dtype as far as I know. If I only pass in an uninitialized `Tensor result;` I'm not aware of any way to convert the `result` to a different dtype with in-place methods.